docs/tutorials/time-series.md

```diff
@@ -0,0 +1,738 @@
+
+  * It takes advantage of the fact that timestamps in a time series grow monotonically or
+    almost monotonically to discard old data no longer needed to compute outputs
+    for new data points. In order to perform these optimizations the query engine needs
+    to know which columns in the input time series grow monotonically.  This requires
+    [`LATENESS` annotations](#timestamp-columns-and-lateness), described below.
+
+  * Time series data is often append-only: once an event has been appended to the
+    series, it cannot be deleted or modified.  Feldera implements an optimized version
+    of several operators for append-only collections.  In order to enable these
+    optimizations, the query engine needs to know which inputs are append-only.  The user
+    can provide this information to the engine using [`append_only` annotatuons](#append-only-tables).
+
+As a user, you do not need to understand the intricate details of time series
+processing in order to take advantage of these optimizations. Your simply
+communicate information about the input data to the query engine using
+annotations.  Feldera uses this information to apply optimizations automatically
+in a way that does not affect the output of the queries.  **Assuming that user
+annotations are correct, the optimized query plan produces the same outputs as
+the unoptimized plan.**
+
+**Users can control the timing of the output of time series queries**.  Users
+can further take advantage of the monotonicity of time series inputs to control
+**when** the output of a query is produced by Feldera using
+[`emit_final`](#emitting-final-values-of-a-view-with-emit_final)
+annotations.
+
+In the rest of this guide, we discuss these concepts in more detail.
+
+## Timestamp columns and lateness
+
+A timestamp column is a column in a SQL table or view that indicates a specific
+point in time associated with each row.
+
+  * There can be any number of timestamp columns in a table. For instance, a
+    table that describes taxi rides can store the time when the ride order was
+    received, pickup time, dropoff time, and the time when the payment was
+    processed.
+
+  * Timestamp columns can have any temporal (`TIMESTAMP`, `DATE`, `TIME`) or
+    numeric type.
+
+**A timestamp column can grow monotonically or almost monotonically**. Time
+series data often arrives ordered by timestamp, i.e., every event has the same
+or larger timestamp than the previous event.  In some cases, events can get
+reordered and delayed, but this delay is bounded, e.g., it may not exceed 1
+hour.  We refer to this bound as **lateness**.
+
+**Lateness** is a constant bound `L` associated with a timestamp column in a
+table or view, such that if the table contains a record with timestamp equal to
+`TS`, then any future records inserted to or removed from the table will have
+timestamps greater than or equal to `TS - L`.  In other words, updates to the
```

Comment:
Sorry, not sure what you're suggesting.

* I like the notion of stream time. It's very simple and helps to explain things, so I don't think we should drop that
* MAX(TS) isn't exactly correct, because of deletions.
* Waterline is only equal to `TS - L` for tables, views are more complicated, therefore I don't really explain how we compute these waterlines, just say that they exist. We can maybe add an appendix or a separate doc explaining: (1) rules for propagating waterlines, (2) rules for doing GC.